### PR TITLE
[RA-5413] Debian 11 crashes after the kernel update

### PR DESCRIPTION
**Pull request by stanislavbarantsev**
The elastio-snap driver uses `System-<ver>.map` file to obtain addresses for some unexported
[and hence normally unavailable] kernel functions. Starting from Debian 11, for some architectures,
this file is available in `linux-image-<ver>-dbg` package. This is a known thing, but it was previously
the case for ARM-based distros only - until this happened on x86. Because of that, no correct
`System.map` is available during the package upgrade, causing the fallback to `/proc/kallsyms` and 
incorrect symbol addresses for the new kernel. This, eventually, causes the driver to crash after reboot.

As the debug package is absolutely huge, it is suggested to download the deb package without
installation and to extract the System.map only. This PR introduces this fix.

### CODE REVIEW

File: src/genconfig.sh [hunk 1]
```diff
@@ -32,12 +54,21 @@ if [ ! -f "$SYSTEM_MAP_FILE" ] || [ $(cat "$SYSTEM_MAP_FILE" | wc -l) -lt 10 ];
...
-		SYSTEM_MAP_FILE="/proc/kallsyms"
-		if [ "$EUID" -ne 0 ]; then
-			echo "Run 'make' command as sudo or root. Otherwise it is not possible to get addresses from the $SYSTEM_MAP_FILE"
-			exit 1
+		# Obtain the relevant System.map file from the
+		# dbg package if the package is being upgraded
+		if [ "$(uname -r)" != "$KERNEL_VERSION" ]; then
+			echo "No System.map found, trying to extract it from the *.deb package"
+			if [ -f /etc/debian_version ] && ! extract_system_map; then
+				exit 1
```
Comment (kgermanov) [inline src/genconfig.sh:65]: Lets add error message about this

File: src/genconfig.sh [hunk 2]
```diff
@@ -23,6 +23,28 @@ if [ ! -z "$1" ]; then
...
 fi
 
+# As a fallback mechanism, if System.map is not found, download
+# the debug linux kernel package and extract it from there
+extract_system_map() {
+	LINUX_IMAGE_DBG="linux-image-$KERNEL_VERSION-dbg"
+	URL=$(sudo apt-get download --print-uris linux-image-$KERNEL_VERSION-dbg | awk -F\' {'print $2'})
+	echo "Downloading $LINUX_IMAGE_DBG from $URL..."
+	if ! wget -q "$URL"; then
+		return 1
```
Comment (kgermanov) [inline src/genconfig.sh:34]: Lets add error message about this

File: src/genconfig.sh [hunk 3]
```diff
@@ -23,6 +23,28 @@ if [ ! -z "$1" ]; then
...
+	fi
+
+	echo "Unpacking..."
+	ar x linux-image-$KERNEL_VERSION-dbg*.deb
```
Comment (kgermanov) [inline src/genconfig.sh]: ar x linux-image-$KERNEL_VERSION-dbg*.deb || return 2

File: src/genconfig.sh [hunk 4]
```diff
@@ -23,6 +23,28 @@ if [ ! -z "$1" ]; then
...
+	rm -f control.tar.xz linux-image-$KERNEL_VERSION-dbg*.deb debian-binary
+
+	echo "Processing. This may take a while..."
+	tar -xvf data.tar.xz -C / "./usr/lib/debug/boot/System.map-$KERNEL_VERSION"
```
Comment (kgermanov) [inline src/genconfig.sh]: tar -xvf data.tar.xz -C / "./usr/lib/debug/boot/System.map-$KERNEL_VERSION" || return 3

File: src/genconfig.sh [hunk 5]
```diff
@@ -23,6 +23,28 @@ if [ ! -z "$1" ]; then
 	KERNEL_VERSION="$1"
 fi
 
+# As a fallback mechanism, if System.map is not found, download
+# the debug linux kernel package and extract it from there
+extract_system_map() {
```
Comment (kgermanov) [inline src/genconfig.sh]: May be rename extract_system_map_for_deb?

**Review by kgermanov — CHANGES_REQUESTED** [review]

File: src/genconfig.sh (hunk 5 above)
Comment (stanislavbarantsev) [reply src/genconfig.sh]: Done.

File: src/genconfig.sh (hunk 4 above)
Comment (stanislavbarantsev) [reply src/genconfig.sh]: Done.

File: src/genconfig.sh (hunk 3 above)
Comment (stanislavbarantsev) [reply src/genconfig.sh]: Done.

File: src/genconfig.sh (hunk 2 above)
Comment (stanislavbarantsev) [reply src/genconfig.sh:34]: Done.

File: src/genconfig.sh (hunk 1 above)
Comment (stanislavbarantsev) [reply src/genconfig.sh:65]: Done.

**Review by kgermanov — APPROVED** [review]

**Review by AnatolyZ — APPROVED** [review]

**Pull request merged**